### ISSUE TITLE
Exception when Filtering Reads

### ISSUE THREAD

**Issue by imk1** [ENCODE-DCC/atac-seq-pipeline#142]
**Describe the bug**
I have been getting an exception that seems to be related to multiprocessing in the step for filtering reads.  I tried pulling the latest version of the pipeline, uninstalling dependencies, reinstalling dependencies, re-configuring the environment, and updating the environment, but that did not help.  Thanks in advance for helping me figure this out!

**OS/Platform and dependencies**
- OS or Platform: Linux cluster with Centos 6 and slurm
- Conda version: If you have used Conda (`$ conda --version`). 4.6.14
- Caper version: N/A
- If not using Caper
  - Cromwell/dxWDL version: cromwell-34.jar

**Attach error logs**
Caper users:
Caper automatically run troubleshooter for failed workflows. If not and If you ran a Caper server and then get workflow_id of your failed workflow with `caper list`. Or directory use a `metadata.json` file on Cromwell's output directory
```
$ caper troubleshoot [WORKFLOW_ID_OR_METADATA_JSON_FILE]
```

Non-Caper users:
1) Move to your working directory where you ran a pipeline. You should be able to find a directory named `cromwell-executions/` which includes all outputs and logs for debugging.

2) Run the following command line to print all non-empty STDERR outputs. This will be greatly helpful for developers to figure out the problem. Copy-paste its output to the issue page.
```
$ find -name stderr -not -empty | xargs tail -n +1
```
==> ./call-filter/shard-0/execution/stderr <==
Traceback (most recent call last):
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/bin/encode_filter.py", line 391, in <module>
    main()
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/bin/encode_filter.py", line 373, in main
    pbc_qc = ret_val_3.get(BIG_INT)
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/lib/python2.7/multiprocessing/pool.py", line 567, in get
    raise self._value
Exception: PID=190834, PGID=190834, RC=1
STDERR=
STDOUT=

==> ./call-filter/shard-1/execution/stderr <==
Traceback (most recent call last):
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/bin/encode_filter.py", line 391, in <module>
    main()
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/bin/encode_filter.py", line 373, in main
    pbc_qc = ret_val_3.get(BIG_INT)
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/lib/python2.7/multiprocessing/pool.py", line 567, in get
    raise self._value
Exception: PID=195411, PGID=195411, RC=1
STDERR=
STDOUT=

3) (OPTIONAL) Run the following command to collect all logs. For developer's convenience, please add `[ISSUE_ID]` to the name of the tar ball file. This command will generate a tar ball including all debugging information. Post an issue with the tar ball (`.tar.gz`) attached.
```
$ find . -type f -name 'stdout' -or -name 'stderr' -or -name 'script' -or \
-name '*.qc' -or -name '*.txt' -or -name '*.log' -or -name '*.png' -or -name '*.pdf' \
| xargs tar -zcvf debug_[ISSUE_ID].tar.gz
```

**Comment by leepc12** [comment]
We have some issues with python multprocessing on SLURM (or SLURM + singularilty). python multiprocessing will be removed in the next release (this week or next week). I think all py-multiprocessing related problems will be fixed in the next release.

Please post your `debug.tar.gz`.

Next time please use Caper, which automatically catch/show errors and help troubleshooting it.

**Comment by imk1** [comment]
Thanks for helping me figure this out.

I re-ran the pipeline with caper and got the same error.  Here is the output from "caper troubleshoot metadata.json:"

Found failures:
[
    {
        "message": "Workflow failed",
        "causedBy": [
            {
                "message": "Job atac.filter:0:2 exited with return code 1 which has not been declared as a valid return code. See 'continueOnReturnCode' runtime attribute for more details.",
                "causedBy": []
            },
            {
                "message": "Job atac.filter:1:2 exited with return code 1 which has not been declared as a valid return code. See 'continueOnReturnCode' runtime attribute for more details.",
                "causedBy": []
            }
        ]
    }
]

atac.filter RetryableFailure. SHARD_IDX=0, RC=1, JOB_ID=114697, RUN_START=2019-07-26T12:36:44.891Z, RUN_END=2019-07-26T15:04:49.725Z, STDOUT=/projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseDNase/LiverFemaleLiu_ATAC_out/atac/f74fa765-bc99-47cf-8eb8-838c8ad7606b/call-filter/shard-0/execution/stdout, STDERR=/projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseDNase/LiverFemaleLiu_ATAC_out/atac/f74fa765-bc99-47cf-8eb8-838c8ad7606b/call-filter/shard-0/execution/stderr
STDERR_CONTENTS=
Traceback (most recent call last):
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/bin/encode_filter.py", line 391, in <module>
    main()
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/bin/encode_filter.py", line 373, in main
    pbc_qc = ret_val_3.get(BIG_INT)
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/lib/python2.7/multiprocessing/pool.py", line 567, in get
    raise self._value
Exception: PID=129360, PGID=129360, RC=1
STDERR=
STDOUT=


atac.filter Failed. SHARD_IDX=0, RC=1, JOB_ID=129732, RUN_START=2019-07-26T15:04:50.888Z, RUN_END=2019-07-26T17:28:46.894Z, STDOUT=/projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseDNase/LiverFemaleLiu_ATAC_out/atac/f74fa765-bc99-47cf-8eb8-838c8ad7606b/call-filter/shard-0/attempt-2/execution/stdout, STDERR=/projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseDNase/LiverFemaleLiu_ATAC_out/atac/f74fa765-bc99-47cf-8eb8-838c8ad7606b/call-filter/shard-0/attempt-2/execution/stderr
STDERR_CONTENTS=
Traceback (most recent call last):
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/bin/encode_filter.py", line 391, in <module>
    main()
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/bin/encode_filter.py", line 373, in main
    pbc_qc = ret_val_3.get(BIG_INT)
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/lib/python2.7/multiprocessing/pool.py", line 567, in get
    raise self._value
Exception: PID=144179, PGID=144179, RC=1
STDERR=
STDOUT=


atac.filter RetryableFailure. SHARD_IDX=1, RC=1, JOB_ID=116959, RUN_START=2019-07-26T12:58:40.890Z, RUN_END=2019-07-26T15:18:24.723Z, STDOUT=/projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseDNase/LiverFemaleLiu_ATAC_out/atac/f74fa765-bc99-47cf-8eb8-838c8ad7606b/call-filter/shard-1/execution/stdout, STDERR=/projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseDNase/LiverFemaleLiu_ATAC_out/atac/f74fa765-bc99-47cf-8eb8-838c8ad7606b/call-filter/shard-1/execution/stderr
STDERR_CONTENTS=
Traceback (most recent call last):
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/bin/encode_filter.py", line 391, in <module>
    main()
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/bin/encode_filter.py", line 373, in main
    pbc_qc = ret_val_3.get(BIG_INT)
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/lib/python2.7/multiprocessing/pool.py", line 567, in get
    raise self._value
Exception: PID=130810, PGID=130810, RC=1
STDERR=
STDOUT=


atac.filter Failed. SHARD_IDX=1, RC=1, JOB_ID=131272, RUN_START=2019-07-26T15:18:26.895Z, RUN_END=2019-07-26T17:36:37.772Z, STDOUT=/projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseDNase/LiverFemaleLiu_ATAC_out/atac/f74fa765-bc99-47cf-8eb8-838c8ad7606b/call-filter/shard-1/attempt-2/execution/stdout, STDERR=/projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseDNase/LiverFemaleLiu_ATAC_out/atac/f74fa765-bc99-47cf-8eb8-838c8ad7606b/call-filter/shard-1/attempt-2/execution/stderr
STDERR_CONTENTS=
Traceback (most recent call last):
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/bin/encode_filter.py", line 391, in <module>
    main()
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/bin/encode_filter.py", line 373, in main
    pbc_qc = ret_val_3.get(BIG_INT)
  File "/home/ikaplow/anaconda2/envs/encode-atac-seq-pipeline/lib/python2.7/multiprocessing/pool.py", line 567, in get
    raise self._value
Exception: PID=144695, PGID=144695, RC=1
STDERR=
STDOUT=

Here is the debug tarred directory:
[debug_142.tar.gz](https://github.com/ENCODE-DCC/atac-seq-pipeline/files/3436901/debug_142.tar.gz)

**Comment by leepc12** [comment]
It's hard to figure out but it looks related to `sambamba` or `python`, `multiprocessing`. We will remove both in the next release.

**Comment by imk1** [comment]
I finally figured out what the problem was.  I ran the filtering part of the pipeline step-by-step and discovered that almost all of the reads got removed when I filtered reads whose ends mapped to different chromosomes.  After some e-mail exchanges with the people who produced the dataset, I figured out that the file that was labeled end 2 was actually technical replicate 2, and the files with the 2nd end of the reads had never been put on SRA.  So it seems like the filtering step of the pipeline is working properly after all.

I am sorry for posting a bug that was not actually a bug.

In future versions of the pipeline, if almost all of the reads are removed in a filtering step, it would be great if some error message could be printed indicating this to help with debugging.  Thanks!
